# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: security/ct/tests/gtest/MultiLogCTVerifierTest.cpp
# Commit: dc509eab3ed7
# Full Hash: dc509eab3ed7a513258723eb46b25f33a610e720
# Author: Dana Keeler <dkeeler@mozilla.com>
# Date: 2025-07-11 16:36:21
# Regressor Bug: 1953221
# File Overlap Count: 3
# Description:
#   Bug 1953221 - don't accept SCTs with a timestamp after the distrustAfter date for the root r=jschanck
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D256720
# ==============================================================================

diff -r 5b86ae2c81f9 -r dc509eab3ed7 security/ct/tests/gtest/MultiLogCTVerifierTest.cpp
--- a/security/ct/tests/gtest/MultiLogCTVerifierTest.cpp	Thu Jul 10 18:37:49 2025 +0000
+++ b/security/ct/tests/gtest/MultiLogCTVerifierTest.cpp	Thu Jul 10 19:04:10 2025 +0000
@@ -82,7 +82,7 @@
     ASSERT_EQ(Success,
               mVerifier.Verify(InputForBuffer(cert), InputForBuffer(issuerSPKI),
                                InputForBuffer(sctList), Input(), Input(), mNow,
-                               result));
+                               Nothing(), result));
     CheckForSingleValidSCTInResult(result, SCTOrigin::Embedded);
   }
 
@@ -153,9 +153,9 @@
   EncodeSCTListForTesting(InputForBuffer(sct), sctList);
 
   CTVerifyResult result;
-  ASSERT_EQ(Success,
-            mVerifier.Verify(InputForBuffer(mTestCert), Input(), Input(),
-                             InputForBuffer(sctList), Input(), mNow, result));
+  ASSERT_EQ(Success, mVerifier.Verify(InputForBuffer(mTestCert), Input(),
+                                      Input(), InputForBuffer(sctList), Input(),
+                                      mNow, Nothing(), result));
 
   CheckForSingleValidSCTInResult(result, SCTOrigin::OCSPResponse);
 }
@@ -166,9 +166,9 @@
   EncodeSCTListForTesting(InputForBuffer(sct), sctList);
 
   CTVerifyResult result;
-  ASSERT_EQ(Success,
-            mVerifier.Verify(InputForBuffer(mTestCert), Input(), Input(),
-                             Input(), InputForBuffer(sctList), mNow, result));
+  ASSERT_EQ(Success, mVerifier.Verify(InputForBuffer(mTestCert), Input(),
+                                      Input(), Input(), InputForBuffer(sctList),
+                                      mNow, Nothing(), result));
 
   CheckForSingleValidSCTInResult(result, SCTOrigin::TLSExtension);
 }
@@ -179,9 +179,10 @@
   EncodeSCTListForTesting(InputForBuffer(sct), sctList);
 
   CTVerifyResult result;
-  ASSERT_EQ(Success, mVerifier.Verify(InputForBuffer(mTestCert), Input(),
-                                      Input(), InputForBuffer(sctList),
-                                      InputForBuffer(sctList), mNow, result));
+  ASSERT_EQ(Success,
+            mVerifier.Verify(InputForBuffer(mTestCert), Input(), Input(),
+                             InputForBuffer(sctList), InputForBuffer(sctList),
+                             mNow, Nothing(), result));
 
   // The result should contain verified SCTs from TLS and OCSP origins.
   size_t embeddedCount = 0;
@@ -211,9 +212,9 @@
   GetSCTListWithInvalidLogID(sctList);
 
   CTVerifyResult result;
-  ASSERT_EQ(Success,
-            mVerifier.Verify(InputForBuffer(mTestCert), Input(), Input(),
-                             Input(), InputForBuffer(sctList), mNow, result));
+  ASSERT_EQ(Success, mVerifier.Verify(InputForBuffer(mTestCert), Input(),
+                                      Input(), Input(), InputForBuffer(sctList),
+                                      mNow, Nothing(), result));
 
   EXPECT_EQ(0U, result.decodingErrors);
   EXPECT_EQ(0U, result.verifiedScts.size());
@@ -233,9 +234,9 @@
   EncodeSCTListForTesting(InputForBuffer(sct), sctList);
 
   CTVerifyResult result;
-  ASSERT_EQ(Success,
-            verifier.Verify(InputForBuffer(mTestCert), Input(), Input(),
-                            Input(), InputForBuffer(sctList), mNow, result));
+  ASSERT_EQ(Success, verifier.Verify(InputForBuffer(mTestCert), Input(),
+                                     Input(), Input(), InputForBuffer(sctList),
+                                     mNow, Nothing(), result));
 
   EXPECT_EQ(0U, result.decodingErrors);
   ASSERT_EQ(1U, result.verifiedScts.size());
@@ -244,5 +245,36 @@
   EXPECT_EQ(mLogOperatorID, result.verifiedScts[0].logOperatorId);
 }
 
+TEST_F(MultiLogCTVerifierTest, VerifiesSCTFromBeforeDistrust) {
+  Buffer sct(GetTestSignedCertificateTimestamp());
+  Buffer sctList;
+  EncodeSCTListForTesting(InputForBuffer(sct), sctList);
+
+  CTVerifyResult result;
+  ASSERT_EQ(Success, mVerifier.Verify(InputForBuffer(mTestCert), Input(),
+                                      Input(), Input(), InputForBuffer(sctList),
+                                      mNow, Some(mNow), result));
+
+  EXPECT_EQ(0U, result.decodingErrors);
+  EXPECT_EQ(1U, result.verifiedScts.size());
+  EXPECT_EQ(0U, result.sctsWithDistrustedTimestamps);
+}
+
+TEST_F(MultiLogCTVerifierTest, IdentifiesSCTFromAfterDistrust) {
+  Buffer sct(GetTestSignedCertificateTimestamp());
+  Buffer sctList;
+  EncodeSCTListForTesting(InputForBuffer(sct), sctList);
+
+  CTVerifyResult result;
+  ASSERT_EQ(Success,
+            mVerifier.Verify(InputForBuffer(mTestCert), Input(), Input(),
+                             Input(), InputForBuffer(sctList), mNow,
+                             Some(TimeFromEpochInSeconds(100)), result));
+
+  EXPECT_EQ(0U, result.decodingErrors);
+  EXPECT_EQ(0U, result.verifiedScts.size());
+  EXPECT_EQ(1U, result.sctsWithDistrustedTimestamps);
+}
+
 }  // namespace ct
 }  // namespace mozilla